File: app/src/lib/server/fs/vault.ts
import { readdir, readFile, writeFile, unlink, mkdir, stat } from "node:fs/promises";
import { resolve, relative, join, dirname, extname, basename } from "node:path";
import { createHash } from "node:crypto";
import matter from "gray-matter";
import { ulid } from "ulid";
import type { NoteFrontmatter } from "@prometheus/core";
import { VAULT_PATH } from "../env.js";

export interface VaultNote {
  path: string;
  title: string;
  content: string;
  rawContent: string;
  frontmatter: NoteFrontmatter;
  checksum: string;
}

export interface VaultFileEntry {
  name: string;
  path: string;
  isDirectory: boolean;
  children?: VaultFileEntry[];
}

function computeChecksum(content: string): string {
  return createHash("sha256").update(content).digest("hex");
}

function ensureFrontmatter(
  rawContent: string,
  filePath: string,
): { content: string; frontmatter: NoteFrontmatter; body: string } {
  const parsed = matter(rawContent);
  const now = new Date().toISOString();

  const frontmatter: NoteFrontmatter = {
    id: parsed.data.id ?? ulid(),
    title:
      parsed.data.title ??
      basename(filePath, extname(filePath)),
    created: parsed.data.created ?? now,
    modified: parsed.data.modified ?? now,
    tags: parsed.data.tags ?? [],
    theme: parsed.data.theme ?? undefined,
  };

  return {
    content: parsed.content,
    frontmatter,
    body: parsed.content,
  };
}

function serializeNote(frontmatter: NoteFrontmatter, body: string): string {
  // Remove undefined/null values — gray-matter's YAML dumper cannot handle them
  const clean: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(frontmatter)) {
    if (value !== undefined && value !== null) {
      clean[key] = value;
    }
  }
  return matter.stringify(body, clean);
}

export async function getVaultPath(): Promise<string> {
  await mkdir(VAULT_PATH, { recursive: true });
  return VAULT_PATH;
}

export async function readNote(notePath: string): Promise<VaultNote> {
  const vaultPath = await getVaultPath();
  const fullPath = resolve(vaultPath, notePath);
  const rawContent = await readFile(fullPath, "utf-8");
  const checksum = computeChecksum(rawContent);
  const { content, frontmatter } = ensureFrontmatter(rawContent, notePath);

  return {
    path: notePath,
    title: frontmatter.title,
    content,
    rawContent,
    frontmatter,
    checksum,
  };
}

export async function writeNote(
  notePath: string,
  body: string,
  frontmatter: NoteFrontmatter,
): Promise<VaultNote> {
  const vaultPath = await getVaultPath();
  const fullPath = resolve(vaultPath, notePath);

  frontmatter.modified = new Date().toISOString();

  const rawContent = serializeNote(frontmatter, body);
  await mkdir(dirname(fullPath), { recursive: true });
  await writeFile(fullPath, rawContent, "utf-8");

  const checksum = computeChecksum(rawContent);

  return {
    path: notePath,
    title: frontmatter.title,
    content: body,
    rawContent,
    frontmatter,
    checksum,
  };
}

export async function createNote(
  notePath: string,
  title?: string,
): Promise<VaultNote> {
  const noteTitle = title ?? basename(notePath, extname(notePath));
  const now = new Date().toISOString();

  const frontmatter: NoteFrontmatter = {
    id: ulid(),
    title: noteTitle,
    created: now,
    modified: now,
    tags: [],
  };

  const body = `\n# ${noteTitle}\n\n`;

  return writeNote(notePath, body, frontmatter);
}

export async function deleteNote(notePath: string): Promise<void> {
  const vaultPath = await getVaultPath();
  const fullPath = resolve(vaultPath, notePath);
  await unlink(fullPath);
}

export async function listNotes(dirPath: string = ""): Promise<string[]> {
  const vaultPath = await getVaultPath();
  const targetPath = resolve(vaultPath, dirPath);
  const notes: string[] = [];

  async function walk(dir: string) {
    let entries;
    try {
      entries = await readdir(dir, { withFileTypes: true });
    } catch {
      return;
    }

    for (const entry of entries) {
      if (entry.name.startsWith(".")) continue;

      const fullPath = join(dir, entry.name);
      if (entry.isDirectory()) {
        await walk(fullPath);
      } else if (extname(entry.name) === ".md") {
        notes.push(relative(vaultPath, fullPath));
      }
    }
  }

  await walk(targetPath);
  return notes.sort();
}

export async function getFileTree(
  dirPath: string = "",
): Promise<VaultFileEntry[]> {
  const vaultPath = await getVaultPath();
  const targetPath = resolve(vaultPath, dirPath);

  let entries;
  try {
    entries = await readdir(targetPath, { withFileTypes: true });
  } catch {
    return [];
  }

  const result: VaultFileEntry[] = [];

  for (const entry of entries) {
    if (entry.name.startsWith(".")) continue;

    const entryRelPath = dirPath ? join(dirPath, entry.name) : entry.name;

    if (entry.isDirectory()) {
      const children = await getFileTree(entryRelPath);
      result.push({
        name: entry.name,
        path: entryRelPath,
        isDirectory: true,
        children,
      });
    } else if (extname(entry.name) === ".md") {
      result.push({
        name: entry.name,
        path: entryRelPath,
        isDirectory: false,
      });
    }
  }

  return result.sort((a, b) => {
    if (a.isDirectory && !b.isDirectory) return -1;
    if (!a.isDirectory && b.isDirectory) return 1;
    return a.name.localeCompare(b.name);
  });
}

export async function noteExists(notePath: string): Promise<boolean> {
  const vaultPath = await getVaultPath();
  const fullPath = resolve(vaultPath, notePath);
  try {
    await stat(fullPath);
    return true;
  } catch {
    return false;
  }
}

File: internal/service/vault.go
package service

import (
	"crypto/sha256"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/oklog/ulid/v2"
	"gopkg.in/yaml.v3"
)

type NoteFrontmatter struct {
	ID       string   `yaml:"id" json:"id"`
	Title    string   `yaml:"title" json:"title"`
	Created  string   `yaml:"created" json:"created"`
	Modified string   `yaml:"modified" json:"modified"`
	Tags     []string `yaml:"tags" json:"tags"`
	Theme    *string  `yaml:"theme,omitempty" json:"theme"`
}

type VaultNote struct {
	Path        string          `json:"path"`
	Title       string          `json:"title"`
	Content     string          `json:"content"`
	RawContent  string          `json:"-"`
	Frontmatter NoteFrontmatter `json:"frontmatter"`
	Checksum    string          `json:"checksum"`
}

type FileEntry struct {
	Name        string       `json:"name"`
	Path        string       `json:"path"`
	IsDirectory bool         `json:"isDirectory"`
	Children    []*FileEntry `json:"children,omitempty"`
}

type Vault struct {
	path string
}

func NewVault(vaultPath string) *Vault {
	os.MkdirAll(vaultPath, 0o755)
	return &Vault{path: vaultPath}
}

func (v *Vault) Path() string {
	return v.path
}

func computeChecksum(data []byte) string {
	h := sha256.Sum256(data)
	return fmt.Sprintf("%x", h)
}

func newULID() string {
	return ulid.Make().String()
}

// ParseFrontmatterPublic is a public wrapper around parseFrontmatter.
func ParseFrontmatterPublic(raw []byte) (NoteFrontmatter, string) {
	return parseFrontmatter(raw)
}

func parseFrontmatter(raw []byte) (NoteFrontmatter, string) {
	content := string(raw)

	if !strings.HasPrefix(content, "---\n") {
		return NoteFrontmatter{}, content
	}

	end := strings.Index(content[4:], "\n---")
	if end < 0 {
		return NoteFrontmatter{}, content
	}

	fmRaw := content[4 : 4+end]
	body := content[4+end+4:] // skip \n---\n

	var fm NoteFrontmatter
	yaml.Unmarshal([]byte(fmRaw), &fm)

	if fm.Tags == nil {
		fm.Tags = []string{}
	}

	return fm, body
}

func serializeFrontmatter(fm NoteFrontmatter) string {
	// Build YAML manually to control output
	var b strings.Builder
	b.WriteString("---\n")
	b.WriteString(fmt.Sprintf("id: \"%s\"\n", fm.ID))
	b.WriteString(fmt.Sprintf("title: \"%s\"\n", fm.Title))
	b.WriteString(fmt.Sprintf("created: \"%s\"\n", fm.Created))
	b.WriteString(fmt.Sprintf("modified: \"%s\"\n", fm.Modified))

	if len(fm.Tags) > 0 {
		tags := make([]string, len(fm.Tags))
		for i, t := range fm.Tags {
			tags[i] = t
		}
		b.WriteString(fmt.Sprintf("tags: [%s]\n", strings.Join(tags, ", ")))
	} else {
		b.WriteString("tags: []\n")
	}

	if fm.Theme != nil && *fm.Theme != "" {
		b.WriteString(fmt.Sprintf("theme: %s\n", *fm.Theme))
	}

	b.WriteString("---\n")
	return b.String()
}

func (v *Vault) ReadNote(notePath string) (*VaultNote, error) {
	fullPath := filepath.Join(v.path, notePath)
	raw, err := os.ReadFile(fullPath)
	if err != nil {
		return nil, err
	}

	checksum := computeChecksum(raw)
	fm, body := parseFrontmatter(raw)

	if fm.ID == "" {
		fm.ID = newULID()
	}
	if fm.Title == "" {
		fm.Title = strings.TrimSuffix(filepath.Base(notePath), ".md")
	}
	now := time.Now().UTC().Format(time.RFC3339)
	if fm.Created == "" {
		fm.Created = now
	}
	if fm.Modified == "" {
		fm.Modified = now
	}

	return &VaultNote{
		Path:        notePath,
		Title:       fm.Title,
		Content:     body,
		RawContent:  string(raw),
		Frontmatter: fm,
		Checksum:    checksum,
	}, nil
}

func (v *Vault) WriteNote(notePath, body string, fm NoteFrontmatter) (*VaultNote, error) {
	fullPath := filepath.Join(v.path, notePath)
	os.MkdirAll(filepath.Dir(fullPath), 0o755)

	fm.Modified = time.Now().UTC().Format(time.RFC3339)

	rawContent := serializeFrontmatter(fm) + body
	if err := os.WriteFile(fullPath, []byte(rawContent), 0o644); err != nil {
		return nil, err
	}

	checksum := computeChecksum([]byte(rawContent))
	return &VaultNote{
		Path:        notePath,
		Title:       fm.Title,
		Content:     body,
		RawContent:  rawContent,
		Frontmatter: fm,
		Checksum:    checksum,
	}, nil
}

func (v *Vault) CreateNote(notePath string, title string) (*VaultNote, error) {
	if title == "" {
		title = strings.TrimSuffix(filepath.Base(notePath), ".md")
	}
	now := time.Now().UTC().Format(time.RFC3339)
	fm := NoteFrontmatter{
		ID:       newULID(),
		Title:    title,
		Created:  now,
		Modified: now,
		Tags:     []string{},
	}
	body := "\n# " + title + "\n\n"
	return v.WriteNote(notePath, body, fm)
}

func (v *Vault) DeleteNote(notePath string) error {
	return os.Remove(filepath.Join(v.path, notePath))
}

func (v *Vault) NoteExists(notePath string) bool {
	_, err := os.Stat(filepath.Join(v.path, notePath))
	return err == nil
}

// isHiddenNote returns true for config files that should not appear in the UI.
func isHiddenNote(name string) bool {
	upper := strings.ToUpper(name)
	return upper == "CLAUDE.MD" || upper == "CLAUDE.LOCAL.MD" || upper == "AGENTS.MD" || upper == "AGENTS.LOCAL.MD"
}

func (v *Vault) ListNotes() ([]string, error) {
	var notes []string
	err := filepath.WalkDir(v.path, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if d.IsDir() && strings.HasPrefix(d.Name(), ".") {
			return filepath.SkipDir
		}
		if !d.IsDir() && strings.HasSuffix(d.Name(), ".md") && !strings.HasPrefix(d.Name(), ".") && !isHiddenNote(d.Name()) {
			rel, _ := filepath.Rel(v.path, path)
			notes = append(notes, rel)
		}
		return nil
	})
	sort.Strings(notes)
	return notes, err
}

func (v *Vault) GetFileTree(dirPath string) ([]*FileEntry, error) {
	targetPath := filepath.Join(v.path, dirPath)
	entries, err := os.ReadDir(targetPath)
	if err != nil {
		return nil, err
	}

	var result []*FileEntry
	for _, entry := range entries {
		if strings.HasPrefix(entry.Name(), ".") {
			continue
		}

		entryRelPath := entry.Name()
		if dirPath != "" {
			entryRelPath = filepath.Join(dirPath, entry.Name())
		}

		if entry.IsDir() {
			children, _ := v.GetFileTree(entryRelPath)
			result = append(result, &FileEntry{
				Name:        entry.Name(),
				Path:        entryRelPath,
				IsDirectory: true,
				Children:    children,
			})
		} else if strings.HasSuffix(entry.Name(), ".md") && !isHiddenNote(entry.Name()) {
			result = append(result, &FileEntry{
				Name:        entry.Name(),
				Path:        entryRelPath,
				IsDirectory: false,
			})
		}
	}

	// Sort: directories first, then alphabetically
	sort.Slice(result, func(i, j int) bool {
		if result[i].IsDirectory && !result[j].IsDirectory {
			return true
		}
		if !result[i].IsDirectory && result[j].IsDirectory {
			return false
		}
		return result[i].Name < result[j].Name
	})

	return result, nil
}
